arch/armv7m/tls.c

```diff
@@ -0,0 +1,34 @@
+/*
+ * Phoenix-RTOS
+ *
+ * libphoenix
+ *
+ * tls.c
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Lukasz Leczkowski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <sys/types.h>
+
+
+volatile uintptr_t arm_tls_ptr __attribute__((section("armtls"), used)) = 0;
+
+
+void __aeabi_read_tp(void)
+{
+	/* __aeabi_read_tp must not clobber any registers except r0.
+	 * `push {r1-r3}` must be done first in separate __asm__ call
+	 * so that no instruction before changes these registers.
+	 */
+	__asm__ volatile("push {r1-r3}");
+	__asm__ volatile("mov r0,%[tls];"
+					 "pop {r1-r3};"
+					 :
+					 : [ tls ] "r"(arm_tls_ptr)
+					 :);
+}
```

Comment:
I think that using inline assembly to define this function is not a good solution.
If this function can only use `r0` register, I suggest you write this function in pure assembler.

Reply:
Ok, rewritten in assembler.